Support for hc::abort through HSAQueue callback

gargrahul opened the pull request:
Added abort function. Mapped to __builtin_trap.

Please find attached the HCC directed test report.
[HCC_DirectedTest_Abort.txt](https://github.com/RadeonOpenCompute/hcc/files/1044480/HCC_DirectedTest_Abort.txt)

gargrahul committed:
Support for hc::abort through HSAQueue callback

whchung reviewed (changes requested):
Please help provide a test case. In this PR you are introducing an new API in user visible interface. Please create a new test case which actually use it.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -787,13 +787,17 @@ namespace Kalmar {
 
         /// Create a queue using the maximum size.
-        hsa_status_t status = hsa_queue_create(agent, queue_size, HSA_QUEUE_TYPE_SINGLE, NULL, NULL,
+        hsa_status_t status = hsa_queue_create(agent, queue_size, HSA_QUEUE_TYPE_SINGLE, callbackQueue /*NULL*/, NULL,
```
@gargrahul Please remove `/*NULL*/` here.

scchan commented on include/hc.hpp:
```diff
@@ -3202,7 +3202,9 @@ inline int __mad24(int x, int y, int z) [[hc]] {
 }
 
-
+extern "C" inline void abort() __HC__ {
```
we don't need extern "C" here since this abort function is in the hc namespace

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -787,13 +787,17 @@ namespace Kalmar {
 // points to RocrQueue, or to nullptr if the HSAQueue is not currently attached to a RocrQueue.
 struct RocrQueue {
+    static void callbackQueue(hsa_status_t status, hsa_queue_t* queue, void *data) {
+        STATUS_CHECK(status, __LINE__);
```
Could you described how we could distinguish different abort conditions (i.e. vm fault, user abort(), etc..)?

gargrahul committed:
Update in Pull request

- Removed extern "C" from abort function def
- Added a test case (uses mad24 test with abort call inside kernel)

gargrahul commented:
*@whchung *I have added a test case in 2nd commit.

On Thu, Jun 1, 2017 at 4:27 PM, Wen-Heng (Jack) Chung <
notifications@github.com> wrote:

> *@whchung* requested changes on this pull request.
>
> Please help provide a test case.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RadeonOpenCompute/hcc/pull/348#pullrequestreview-41476265>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD_A9w-RqsKVp9_O5L3M80nxut42tYqVks5r_pkUgaJpZM4Ns0Af>
> .
>

scchan commented on tests/Unit/HC/abort.cpp:
```diff
@@ -0,0 +1,93 @@
+
+// RUN: %hc %s -o %t.out && %t.out
+
```
This test is expected to abort during execution but it would be counted as a test failure if it's run as is.
The proper way would be to pipe stdout and stderr into the FileCheck tool to look for the expected error message.  You'll find examples with FileCheck in hcc/tests/Unit

whchung commented on tests/Unit/HC/abort.cpp:
```diff
@@ -0,0 +1,93 @@
+
+// RUN: %hc %s -o %t.out && %t.out
```
@gargrahul the test is incorrect as @scchan pointed out. Once you create a test you should run it with `make test` to make sure it works.

whchung reviewed (changes requested):
the test is invalid.

gargrahul committed:
Removed abort function test file

scchan commented:
@whchung could you review and approve?
